### PR TITLE
Add `@metadata.input_id` and `@metadata.stream_id` when injecting streams

### PR DESCRIPTION
**Pull request by rdner**
## What does this PR do?

This PR adds 2 additional values added to each event via the inject stream processor.

These 2 value are going to be used in the shipper to identify where an
event came from in order to apply processors accordingly.

Also, added test cases for the processor to verify the change.

**Since I'm not well familiar with the code base, I'm not 100% sure the IDs that I extract will be present before the processor is applied. Also, I didn't find where the IDs could be generated.**

## Why is it important?

It's needed to implement [Input/Stream Identification in the new Shipper Pipeline](https://docs.google.com/document/d/1fI6vBcxpxAhS4YtEhpveKR8y_irqPNqSOW2RTuRX8cc/edit) proposal.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

I ran `go test -run ".*TestRules.*"` in the `internal/pkg/agent/transpiler` package

## Related issues

- Relates https://github.com/elastic/elastic-agent-shipper/issues/24

### CODE REVIEW

**Comment by rdner** [comment]
@cmacknz yes, I agree with everything you wrote.

I'm aware that the upcoming spec V2 will change everything and will require additional work. This is to unblock my current work on the shipper output in beats as a temporary solution. I expect something different for the spec V2 later.

**Comment by cmacknz** [comment]
Added https://github.com/elastic/elastic-agent-shipper/issues/52 to track doing this without relying on specification file transformations as required by the V2 architecture.

File: internal/pkg/agent/transpiler/rules.go [hunk 1]
```diff
@@ -669,6 +669,42 @@ func (r *InjectStreamProcessorRule) Apply(_ AgentInfo, ast *AST) (err error) {
...
+			}
+
+			processorsList, ok := processorsNode.Value().(*List)
+			if !ok {
+				return errors.New("InjectStreamProcessorRule: input processors is not a list")
+			}
+
+			// inject `input_id` on the input level
+			processorMap := &Dict{value: make([]Node, 0)}
+			processorMap.value = append(processorMap.value, &Key{name: "target", value: &StrVal{value: "source"}})
```
Comment (cmacknz) [inline internal/pkg/agent/transpiler/rules.go:700]: I think you are currently adding these fields into the event payload, such that they will be sent to and stored in Elasticsearch eventually. 

This change makes every event sent from agent slightly larger, which I don't think we want. For now they will exclusively be used by the shipper and nothing else.

Let's include the input ID and stream ID in the event metadata instead. Then the gRPC output should be able to extract them and add them to the relevant protobuf message fields without needing to store them in Elasticsearch afterwards.

The `add_fields` documentation explains how to do this https://www.elastic.co/guide/en/beats/filebeat/current/add-fields.html:

```yml
processors:
  - add_fields:
      target: '@metadata'
      fields:
        # ...
```

Comment (rdner) [reply internal/pkg/agent/transpiler/rules.go]: @cmacknz I actually worked on supporting `@metadata` as the `target` parameter in `adds_fields` processor, at last it showed itself useful :)

I've updated everything accordingly.

**Comment by ph** [comment]
Small question here concerning naming, do we really need the **prefix** `input_` or `stream_` we are already in an input and stream section ? wdyt @blakerouse

**Comment by ph** [comment]
@rdner Sound good, I get it now thanks with your details, I was completely out of the loop.

**Comment by michalpristas** [comment]
testing now

File: internal/pkg/agent/transpiler/rules.go [hunk 2]
```diff
@@ -669,6 +669,42 @@ func (r *InjectStreamProcessorRule) Apply(_ AgentInfo, ast *AST) (err error) {
...
+			processorsNode, found := inputNode.Find("processors")
+			if !found {
+				processorsNode = &Key{
+					name:  "processors",
+					value: &List{value: make([]Node, 0)},
+				}
+
+				inputMap, ok := inputNode.(*Dict)
+				if ok {
+					inputMap.value = append(inputMap.value, processorsNode)
```
Comment (michalpristas) [inline internal/pkg/agent/transpiler/rules.go:689]: what if input already has some processor defined

**Review by michalpristas — APPROVED** [review]
🚢 🇮🇹

**Pull request merged**